readline/buffer.go

```diff
@@ -121,114 +165,299 @@ func (b *Buffer) MoveToEnd() {
 			remainder := b.Size() % b.LineWidth
 			fmt.Printf(CursorBOL + cursorRightN(len(b.Prompt.prompt())+remainder))
 		} else {
-			fmt.Print(cursorRightN(b.Size() - b.Pos))
+			fmt.Print(cursorRightN(b.Size() - b.DisplayPos))
 		}
 
-		b.Pos = b.Size()
+		b.Pos = b.Buf.Size()
+		b.DisplayPos = b.Size()
 	}
 }
 
 func (b *Buffer) Size() int {
-	return b.Buf.Size()
+	sum := 0
+	for i := 0; i < b.Buf.Size(); i++ {
+		if e, ok := b.Buf.Get(i); ok {
+			if r, ok := e.(rune); ok {
+				sum += runewidth.RuneWidth(r)
+			}
+		}
+	}
+
+	return sum
 }
 
 func (b *Buffer) Add(r rune) {
+	rLength := runewidth.RuneWidth(r)
+
 	if b.Pos == b.Buf.Size() {
-		fmt.Printf("%c", r)
+
+		b.DisplayPos += rLength
+
+		if b.Pos > 0 {
+
+			if b.DisplayPos%b.LineWidth == 0 {
+				fmt.Printf("%c", r)
+				fmt.Printf("\n%s", b.Prompt.AltPrompt)
+				b.LineFlags.Add(false)
+
+			} else if b.DisplayPos%b.LineWidth < (b.DisplayPos-rLength)%b.LineWidth {
+				fmt.Printf("\n%s", b.Prompt.AltPrompt)
+				b.DisplayPos += 1
+				fmt.Printf("%c", r)
+				b.LineFlags.Add(true)
+
+			} else {
+				fmt.Printf("%c", r)
+			}
+		} else {
+			fmt.Printf("%c", r)
+		}
+
+		b.Buf.Add(r)
+		b.Pos += 1
+
+		/*fmt.Printf("%c", r)
 		b.Buf.Add(r)
 		b.Pos += 1
-		if b.Pos > 0 && b.Pos%b.LineWidth == 0 {
+		b.DisplayPos += rLength
+		if b.Pos > 0 && (b.DisplayPos%b.LineWidth < (b.DisplayPos-rLength)%b.LineWidth || b.DisplayPos%b.LineWidth == 0) {
 			fmt.Printf("\n%s", b.Prompt.AltPrompt)
-		}
+			if b.DisplayPos%b.LineWidth == 1 {
+				b.DisplayPos -= 1
+			}
+		}*/
+
+		//INSERTING
 	} else {
-		fmt.Printf("%c", r)
+
+		b.DisplayPos += rLength
+		if b.Pos > 0 {
+
+			if b.DisplayPos%b.LineWidth == 0 {
+				fmt.Printf("%c", r)
+				fmt.Printf("\n%s", b.Prompt.AltPrompt)
+				b.LineFlags.Set(b.DisplayPos/b.LineWidth-1, false)
+
+			} else if b.DisplayPos%b.LineWidth < (b.DisplayPos-rLength)%b.LineWidth {
+				fmt.Print(ClearToEOL)
+				fmt.Printf("\n%s", b.Prompt.AltPrompt)
+				b.DisplayPos += 1
+				fmt.Printf("%c", r)
+				b.LineFlags.Set(b.DisplayPos/b.LineWidth-1, true)
+
+			} else {
+				fmt.Printf("%c", r)
+			}
+		} else {
+			fmt.Printf("%c", r)
+		}
+		b.Buf.Insert(b.Pos, r)
+		b.Pos += 1
+
+		/*fmt.Printf("%c", r)
 		b.Buf.Insert(b.Pos, r)
 		b.Pos += 1
-		if b.Pos > 0 && b.Pos%b.LineWidth == 0 {
+		b.DisplayPos += rLength
+
+		if b.Pos > 0 && b.DisplayPos%b.LineWidth < (b.DisplayPos-rLength)%b.LineWidth || b.DisplayPos%b.LineWidth == 0 {
 			fmt.Printf("\n%s", b.Prompt.AltPrompt)
-		}
+			if b.DisplayPos%b.LineWidth == 1 {
+				b.DisplayPos -= 1
+			}
+		}*/
+
 		b.drawRemaining()
 	}
 }
 
+func (b *Buffer) countCurrLength(place int) int {
+	var sum int
+	var counter int = -1
+	var prevLen int
+
+	for place <= b.LineWidth {
+
+		counter += 1
+		sum += prevLen
+		if e, ok := b.Buf.Get(b.Pos + counter); ok {
+			if r, ok := e.(rune); ok {
+				place += runewidth.RuneWidth(r)
+				prevLen = len(string(r))
+			}
+		} else {
```

Comment:
think its possible we could get rid of this `else { break }`? Seems possible the next condition could cover this, but Im not sure.